ufl/argument.py

```diff
@@ -163,7 +225,11 @@ def Arguments(function_space, number):
     """UFL value: Create an Argument in a mixed space, and return a
     tuple with the function components corresponding to the subelements."""
     if isinstance(function_space, MixedFunctionSpace):
-        return [Argument(function_space.ufl_sub_space(i), number, i)
+        # return [Argument(function_space.ufl_sub_space(i), number, i)
+        #         for i in range(function_space.num_sub_spaces())]
+        return [Argument(function_space.ufl_sub_spaces()[i], number, i)
+                if is_primal(function_space.ufl_sub_spaces()[i])
+                else Coargument(function_space.ufl_sub_spaces()[i], number, i)
```

Comment:
I guess it makes sense that a `MixedFunctionSpace` can contain a combination of primal and dual spaces. The surface API like this makes me a bit unhappy, in that we've pushed the responsibility for correct construction onto all consumers.

Suggestion: could we have a `def Argument(...)` function that dispatches on `is_primal` and returns either an `Argument` object or a `Coargument` object? I understand why you need to separate the objects, but is it neater if  we say `Argument(dual_space) -> Coargument` and `Argument(primal_space) -> Argument`?

Reply:
The `Argument` constructor already does this, so the code here has been simplified.